Berlin jun2019 presentation and misc fixes

hjorthjort opened the pull request:
The presentation and demo files from Berlin 2019 "Wasm on the blockchain" conference, plus some misc changes to semantics that was useful for the demo.

MrChico commented on wasm.md:
```diff
@@ -254,7 +254,8 @@ Note that we do not need to call `#chop` on the results here.
  // -----------------------------------
-    rule <k> ITYPE . div_u I1 I2 => #if I2 =/=Int 0 #then < ITYPE > (I1 /Int I2) #else undefined #fi ... </k>
+    rule <k> ITYPE . div_u I1 I2 => < ITYPE > (I1 /Int I2)  ... </k> requires I2 =/=Int 0
+    rule <k> ITYPE . div_u I1 I2 => undefined ... </k>               requires I2  ==Int 0
```
we should make a consistent choice on whether to branch on these instructions, in which case  `rem_{u,s}` and `div_s` should be updated as well

ehildenb replied:
Branching (as this does here) is probably better.

hjorthjort replied:
At least it give a lot nicer behavior tree in KLab, which is why I changed it for `div_u` (it's the only one I used in the presentation). I'll change it for the others as well.

ehildenb commented on wasm.md:
```diff
@@ -481,9 +482,10 @@ Note that, unlike in the WebAssembly specification document, we do not need the
 
-    syntax Instr ::= "(" "br_if" Int ")"
+    syntax Instr ::= "(" "br_if" Int ")" | "(" "br_if" Int Instr ")"
  // ------------------------------------
```
Formatting.

hjorthjort replied:
Fixed.

ehildenb commented on wasm.md:
```diff
@@ -542,17 +546,19 @@ The `*_local` instructions are defined here.
-                   | "(" "local.tee" Int ")"
+                   | "(" "local.set" Int ")" | "(" "local.set" Int Instr ")"
+                   | "(" "local.tee" Int ")" | "(" "local.tee" Int Instr ")"
  // ----------------------------------------
```
Formatting.

hjorthjort commented:
@MrChico Would it make sense to merge `oldSchool` to `master` in KLab? I'm currently seeing a crash saying "can't read property `label` of `undefined`" or similar when using the master branch of KLab, guessing this was the fix:

```
-        let k = state.nodes[node.id.split("_")[0]];
-        if(k) {
-          node.oog = kast.get(k, "ethereum.evm.statusCode").label == "EVMC_OUT_OF_GAS_NETWORK"
-          node.gas = kast.get(k, "ethereum.evm.callState.gas")
-        }
+        // let k = state.nodes[node.id.split("_")[0]];
+        // if(k) {
+        //   node.oog = kast.get(k, "ethereum.evm.statusCode").label == "EVMC_OUT_OF_GAS_NETWORK"
+        //   node.gas = kast.get(k, "ethereum.evm.callState.gas")
+        // }
```

Maybe the original code should just be wrapped in some error handling? At least until KLab has a more modular structure with less EVM stuff hard coded.

ehildenb commented on wasm.md:
```diff
@@ -487,14 +492,16 @@ Note that, unlike in the WebAssembly specification document, we do not need the
+    syntax Instr ::= "(" "br_if" Int ")" | "(" "br_if" Int Instr ")"
+ // ----------------------------------------------------------------
+    rule <k> ( br_if N I:Instr ) => I ~> ( br_if N ) ... </k>
  // ------------------------------------
```
formatting here is pretty weird

hjorthjort replied:
Fixed

ehildenb commented on wasm.md:
```diff
@@ -256,26 +256,31 @@ Note that we do not need to call `#chop` on the results here.
-    rule <k> ITYPE . div_u I1 I2 => #if I2 =/=Int 0 #then < ITYPE > (I1 /Int I2) #else undefined #fi ... </k>
-    rule <k> ITYPE . rem_u I1 I2 => #if I2 =/=Int 0 #then < ITYPE > (I1 %Int I2) #else undefined #fi ... </k>
+    rule <k> ITYPE . div_u I1 I2 => < ITYPE > I1 /Int I2 ... </k> requires I2 =/=Int 0
+    rule <k> ITYPE . div_u I1 I2 => undefined ... </k>            requires I2  ==Int 0
```
Can we line up the `...`.

hjorthjort replied:
Done

ehildenb commented on wasm.md:
```diff
@@ -256,26 +256,31 @@ Note that we do not need to call `#chop` on the results here.
+    rule <k> ITYPE . div_u I1 I2 => undefined ... </k>            requires I2  ==Int 0
+
+    rule <k> ITYPE . rem_u I1 I2 => < ITYPE > I1 %Int I2 ... </k> requires I2 =/=Int 0
+    rule <k> ITYPE . rem_u I1 I2 => undefined ... </k>            requires I2  ==Int 0
```
Also here, line up the `...`.

hjorthjort replied:
Done.

hjorthjort commented on wasm.md:
```diff
@@ -256,26 +256,31 @@ Note that we do not need to call `#chop` on the results here.
+          => < ITYPE > #unsigned(ITYPE, #signed(ITYPE, I1) %Int #signed(ITYPE, I2)) ... </k>
+      requires I2 =/=Int 0
+       andBool I1 /Int I2 =/=Int #pow1(ITYPE)
+    rule <k> ITYPE . rem_s I1 I2 => undefined ... </k> requires I2 ==Int 0
```
No, the `andBool` condition shouldn't be there. Good catch!

hjorthjort replied:
Fixed. Also added regression tests.
